netmap/attributes.go

```diff
@@ -0,0 +1,241 @@
+// Attribute Value is:
+//   * `True` if node enters the subnet;
+//   * `False`, otherwise.
+func WriteSubnetInfo(node *NodeInfo, info NodeSubnetInfo) {
+	attrs := node.GetAttributes()
+
+	id := info.ID()
+	enters := info.Enters()
+
+	// calculate attribute key
+	key := subnetAttributeKey(id)
+
+	if refs.IsZeroSubnet(id) == enters {
+		for i := range attrs {
+			if attrs[i].GetKey() == key {
+				attrs = append(attrs[:i], attrs[i+1:]...)
+			}
+		}
+	} else {
+		var val string
+
+		if enters {
+			val = attrSubnetValEntry
+		} else {
+			val = attrSubnetValExit
+		}
+
+		presented := false
+
+		for i := range attrs {
+			if attrs[i].GetKey() == key {
+				attrs[i].SetValue(val)
+				presented = true
+			}
+		}
+
+		if !presented {
+			var attr Attribute
+
+			attr.SetKey(key)
+			attr.SetValue(val)
+
+			attrs = append(attrs, &attr)
+		}
+	}
+
+	node.SetAttributes(attrs)
+}
+
+// ErrRemoveSubnet is returned when a node needs to leave the subnet.
+var ErrRemoveSubnet = errors.New("remove subnet")
+
+var errNoSubnets = errors.New("no subnets")
+
+// IterateSubnets iterates over all subnets the node belongs to and passes the IDs to f.
+// Handler must not be nil.
+//
+// Subnet attributes are expected to be key-unique, otherwise undefined behavior.
+//
+// If f returns ErrRemoveSubnet, then removes subnet entry. Note that this leads to an instant mutation of NodeInfo.
+// Breaks on any other non-nil error and returns it.
+//
+// Returns an error if any subnet attribute has wrong format.
+// Returns an error if the node is not included in any subnet by the end of the loop.
+func IterateSubnets(node *NodeInfo, f func(refs.SubnetID) error) error {
+	attrs := node.GetAttributes()
+
+	type zeroStatus uint8
+
+	const (
+		_ zeroStatus = iota
+		// missing attribute of zero subnet
+		zeroNoAttr
+		// with `False` attribute
+		zeroExit
+		// with `True` attribute
+		zeroEntry
+	)
+
+	var (
+		err     error
+		id      refs.SubnetID
+		entries uint
+
+		stZero = zeroNoAttr
+	)
+
+	for i := 0; i < len(attrs); i++ { // range must not be used because of attrs mutation in body
+		key := attrs[i].GetKey()
+
+		// cut subnet ID string
+		idTxt := strings.TrimPrefix(key, attrSubnetPrefix)
+		if idTxt == key {
+			// not a subnet attribute
+			continue
+		}
+
+		// check value
+		val := attrs[i].GetValue()
+		if val != attrSubnetValExit && val != attrSubnetValEntry {
+			return fmt.Errorf("invalid attribute value: %s", val)
+		}
+
+		// decode subnet ID
+		if err = id.UnmarshalText([]byte(idTxt)); err != nil {
+			return fmt.Errorf("invalid ID text: %w", err)
+		}
+
+		// update status of zero subnet
+		isZero := refs.IsZeroSubnet(&id)
+
+		if stZero == zeroNoAttr { // in order to not reset if has been already set
+			if isZero {
+				if val == attrSubnetValEntry {
+					// clear True attribute for zero subnet is also possible
+					stZero = zeroEntry
+				} else {
+					stZero = zeroExit
+				}
```

Comment:
1. `zeroEntry` and `zeroExit` has only this reference
2. I think single flag should be enough

There is only one case when zero subnetwork should be ignored. 

```go
var ignoreZeroSubnetwork bool
for i := 0; i < len(attrs); i++ {
...
    if isZero {
        ignoreZeroSubnetwork == attrSubnetValExit
    }
...
}
if !ignoreZeroSubnetwork {
    // missing attribute of zero subnet equivalent to entry
    refs.MakeZeroSubnet(&id)
    err = f(id)
    ...
}
```

Reply:
Agree, for now it's enought to distinguish 2 cases.

Reply:
Simplified.